File: samples/bot-conversation-sso-quickstart/csharp_dotnetcore/BotConversationSsoQuickstart/Dialogs/MainDialog.cs
// <copyright file="MainDialog.cs" company="Microsoft">
// Copyright (c) Microsoft. All rights reserved.
// </copyright>

using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Connector.Authentication;
using Microsoft.Bot.Schema;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;

namespace Microsoft.BotBuilderSamples
{
    /// <summary>
    /// Main dialog that handles the authentication and user interactions.
    /// </summary>
    public class MainDialog : LogoutDialog
    {
        protected readonly ILogger _logger;

        /// <summary>
        /// Initializes a new instance of the <see cref="MainDialog"/> class.
        /// </summary>
        /// <param name="configuration">The configuration.</param>
        /// <param name="logger">The logger.</param>
        public MainDialog(IConfiguration configuration, ILogger<MainDialog> logger)
            : base(nameof(MainDialog), configuration["ConnectionName"])
        {
            _logger = logger;

            AddDialog(new OAuthPrompt(
                nameof(OAuthPrompt),
                new OAuthPromptSettings
                {
                    ConnectionName = ConnectionName,
                    Text = "Please Sign In",
                    Title = "Sign In",
                    Timeout = 300000, // User has 5 minutes to login (1000 * 60 * 5)
                    EndOnInvalidMessage = true
                }));

            AddDialog(new ConfirmPrompt(nameof(ConfirmPrompt)));

            AddDialog(new WaterfallDialog(nameof(WaterfallDialog), new WaterfallStep[]
            {
                    PromptStepAsync,
                    LoginStepAsync,
                    DisplayTokenPhase1Async,
                    DisplayTokenPhase2Async,
            }));

            // The initial child Dialog to run.
            InitialDialogId = nameof(WaterfallDialog);
        }

        /// <summary>
        /// Prompts the user to sign in.
        /// </summary>
        /// <param name="stepContext">The waterfall step context.</param>
        /// <param name="cancellationToken">The cancellation token.</param>
        /// <returns>A task representing the asynchronous operation.</returns>
        private async Task<DialogTurnResult> PromptStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            _logger.LogInformation("PromptStepAsync() called.");
            return await stepContext.BeginDialogAsync(nameof(OAuthPrompt), null, cancellationToken);
        }

        /// <summary>
        /// Handles the login step.
        /// </summary>
        /// <param name="stepContext">The waterfall step context.</param>
        /// <param name="cancellationToken">The cancellation token.</param>
        /// <returns>A task representing the asynchronous operation.</returns>
        private async Task<DialogTurnResult> LoginStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var tokenResponse = (TokenResponse)stepContext.Result;
            if (tokenResponse?.Token != null)
            {
                try
                {
                    var client = new SimpleGraphClient(tokenResponse.Token);
                    var me = await client.GetMeAsync();
                    var title = !string.IsNullOrEmpty(me.JobTitle) ? me.JobTitle : "Unknown";

                    await stepContext.Context.SendActivityAsync($"You're logged in as {me.DisplayName} ({me.UserPrincipalName}); your job title is: {title}");

                    var photo = await client.GetPhotoAsync();

                    if (!string.IsNullOrEmpty(photo))
                    {
                        var cardImage = new CardImage(photo);
                        var card = new ThumbnailCard(images: new List<CardImage> { cardImage });
                        var reply = MessageFactory.Attachment(card.ToAttachment());

                        await stepContext.Context.SendActivityAsync(MessageFactory.Text("Sorry! User doesn't have a profile picture to display."), cancellationToken);
                    }
                    else
                    {
                        await stepContext.Context.SendActivityAsync(MessageFactory.Text("Sorry! User doesn't have a profile picture to display."), cancellationToken);
                    }

                    return await stepContext.PromptAsync(
                        nameof(ConfirmPrompt),
                        new PromptOptions { Prompt = MessageFactory.Text("Would you like to view your token?") },
                        cancellationToken);
                }
                catch (Exception ex)
                {
                    _logger.LogError("Error occurred while processing your request.", ex);
                }
            }
            else
            {
                _logger.LogInformation("Response token is null or empty.");
            }

            await stepContext.Context.SendActivityAsync(MessageFactory.Text("Login was not successful, please try again."), cancellationToken);
            return await stepContext.EndDialogAsync(cancellationToken: cancellationToken);
        }

        /// <summary>
        /// Displays the token if the user confirms.
        /// </summary>
        /// <param name="stepContext">The waterfall step context.</param>
        /// <param name="cancellationToken">The cancellation token.</param>
        /// <returns>A task representing the asynchronous operation.</returns>
        private async Task<DialogTurnResult> DisplayTokenPhase1Async(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            _logger.LogInformation("DisplayTokenPhase1Async() method called.");

            await stepContext.Context.SendActivityAsync(MessageFactory.Text("Thank you."), cancellationToken);

            var result = (bool)stepContext.Result;
            if (result)
            {
                return await stepContext.BeginDialogAsync(nameof(OAuthPrompt), cancellationToken: cancellationToken);
            }

            return await stepContext.EndDialogAsync(cancellationToken: cancellationToken);
        }

        /// <summary>
        /// Displays the token to the user.
        /// </summary>
        /// <param name="stepContext">The waterfall step context.</param>
        /// <param name="cancellationToken">The cancellation token.</param>
        /// <returns>A task representing the asynchronous operation.</returns>
        private async Task<DialogTurnResult> DisplayTokenPhase2Async(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            _logger.LogInformation("DisplayTokenPhase2Async() method called.");

            var tokenResponse = (TokenResponse)stepContext.Result;
            if (tokenResponse != null)
            {
                await stepContext.Context.SendActivityAsync(MessageFactory.Text($"Here is your token: {tokenResponse.Token}"), cancellationToken);
            }

            return await stepContext.EndDialogAsync(cancellationToken: cancellationToken);
        }
    }
}

File: samples/TeamsSDK/Archived/bot-teams-authentication/python/dialogs/main_dialog.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

from botbuilder.core import MessageFactory
from botbuilder.dialogs import WaterfallDialog, WaterfallStepContext, DialogTurnResult, PromptOptions
from botbuilder.dialogs.prompts import ConfirmPrompt, OAuthPromptSettings
from custom_oauth_prompt import OAuthPrompt  # Custom wrapper with bug fix
from dialogs import LogoutDialog
import logging

# Set logging
logging.basicConfig(level=logging.INFO)

class MainDialog(LogoutDialog):
    def __init__(self, connection_name: str):
        super(MainDialog, self).__init__(MainDialog.__name__, connection_name)
        self._connection_name = connection_name

        # Add OAuthPrompt for Teams SSO
        self.add_dialog(
            OAuthPrompt(
                OAuthPrompt.__name__,
                OAuthPromptSettings(
                    connection_name=connection_name,
                    text="Please sign in to continue.",
                    title="Sign In",
                    timeout=300000,  # 5 minutes
                ),
            )
        )

        # Add ConfirmPrompt
        self.add_dialog(ConfirmPrompt(ConfirmPrompt.__name__))

        # Waterfall dialog steps
        self.add_dialog(
            WaterfallDialog(
                "WFDialog",
                [
                    self.login_step,
                    self.display_token_phase1,
                    self.display_token_phase2,
                ],
            )
        )

        self.initial_dialog_id = "WFDialog"

    # Step 1: Attempt to get SSO token using OAuthPrompt
    async def login_step(self, step_context: WaterfallStepContext) -> DialogTurnResult:
        # Use OAuthPrompt to get the token (handles SSO and fallback signin)
        return await step_context.begin_dialog(OAuthPrompt.__name__)

    # Step 2: Process the token response
    async def display_token_phase1(self, step_context: WaterfallStepContext) -> DialogTurnResult:
        # The result from OAuthPrompt is the TokenResponse
        token_response = step_context.result
        
        # Check if we got a valid token response (not an error)
        if token_response and hasattr(token_response, 'token') and token_response.token:
            logging.info("Successfully retrieved token via OAuthPrompt")
            # Store token in step values for later use
            step_context.values["token"] = token_response.token
            await step_context.context.send_activity("You are now logged in!")
            return await step_context.prompt(
                ConfirmPrompt.__name__,
                PromptOptions(
                    prompt=MessageFactory.text("Would you like to view your token?")
                ),
            )
        else:
            logging.warning(f"Token response was None or invalid. Response type: {type(token_response)}")
            await step_context.context.send_activity(
                "Login was not successful. Please try again."
            )
            return await step_context.end_dialog()

    # Step 3: Display token to user
    async def display_token_phase2(self, step_context: WaterfallStepContext) -> DialogTurnResult:
        if step_context.result:
            # Retrieve the token from step values
            token = step_context.values.get("token")
            
            if token:
                await step_context.context.send_activity(
                    f"Here is your token: {token}"
                )
            else:
                await step_context.context.send_activity(
                    "Token not available."
                )
        else:
            await step_context.context.send_activity("Thank you.")

        return await step_context.end_dialog()
